streaming/base/stream.py

```diff
@@ -0,0 +1,312 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""A dataset, or sub-dataset if mixing, from which we stream/cache samples."""
+
+import json
+import os
+from tempfile import mkdtemp
+from typing import List, Optional
+
+from typing_extensions import Self
+
+from streaming.base.compression import decompress
+from streaming.base.format import Reader, reader_from_json
+from streaming.base.format.base.reader import FileInfo
+from streaming.base.hashing import get_hash
+from streaming.base.index import get_index_basename
+from streaming.base.storage import download_file
+from streaming.base.util import TICK, wait_for_file_to_exist
+from streaming.base.world import World
+
+
+class Stream:
+    """A dataset, or sub-dataset if mixing, from which we stream/cache samples.
+
+    We initialize a StreamingDataset with one or more Streams. Streams may be resampled to achieve
+    different mixtures of samples.
+
+    Stream init takes three kinds of arguments:
+      * At least one of ``remote`` and ``local`` must exist. If no ``remote``, the data must be
+        local. If no ``local``, we cache to a temp directory.
+          * ``remote``
+          * ``local``
+      * At most one of ``proportion``, ``repeat``, or ``samples`` may exist. If none are provided,
+        each sample is seen once per epoch. If provided one of these, we derive the others.
+        Note that ``proportion`` (relative) and ``repeat``/``samples`` (absolute) are mutually
+        incompatible -- you must entirely use one or the other (or neither) for all sub-datasets.
+          * ``proportion``
+          * ``repeat``
+          * ``samples``
+      * The remaining arguments are optional knobs for controlling downloading behavior and default
+        to ``None``. If ``None``, they take a default value provided to or by the StreamingDataset
+        init.
+          * ``split``
+          * ``download_retry``
+          * ``download_timeout``
+          * ``validate_hash``
+          * ``keep_zip``
+          * ``keep_raw``
+
+    Args:
+        remote (str, optional): Remote path or directory to download the dataset from. If ``None``,
+            its data must exist locally. Defaults to ``None``.
+        local (str, optional): Local working directory to download shards to. This is where shards
+            are cached while they are being used. Uses a temp directory if not set. Defaults to
+            ``None``.
+        split (str, optional): Which dataset split to use, if any. If provided, we stream from/to
+            the ``split`` subdirs of  ``remote`` and ``local``. Defaults to ``None``.
+        proportion (float, optional): How much to upsample or downsample this dataset, as the
```

Comment:
can you rewrite the documentation for this param so it is more clear how it relies on the logic in the encapsulating `StreamingDataset`? i.e. don't use "we" or "any of them"

Reply:
how about now